### PR TITLE
[FLINK-11082][network] Fix the calculation of backlog in PipelinedSubpartition

### PR DESCRIPTION
**Pull request by zhijiangW**
## What is the purpose of the change

*The backlog of subpartition should indicate how many buffers are consumable, then the consumer could feedback the corresponding credits for transporting these buffers. But in current `PipelinedSubpartition` implementation, the backlog is increased by 1 when a `BufferConsumer` is added into `PipelinedSubpartition`, and decreased by 1 when a `BufferConsumer` is removed from `PipelinedSubpartition`.  So the backlog only reflects how many buffers are retained in `PipelinedSubpartition`, which is not always equivalent to the number of consumable buffers.*

*The backlog inconsistency might result in floating buffers misdistribution on consumer side, because the consumer would request floating buffers based on backlog value, then one floating buffer might not be used in `RemoteInputChannel` long time after requesting.*

*Considering the solution, the last buffer in `PipelinedSubpartition` could only be consumable in the case of flush triggered or partition finished. So we could calculate the backlog precisely based on partition flushed/finished conditions.*

## Brief change log

  - *Adjust the implementation in `PipelinedSubpartition#getBuffersInBacklog` to decrease 1 if not flushed or finished.*
  - *Adjust the previous old unit tests and add new unit tests*

## Verifying this change

*Add three new unit tests `testBacklogConsistentWithNumberOfConsumableBuffers` in `PipelinedSubpartitionWithReadViewTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / **don't know**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhijiangW** [comment]
@pnowojski , based on previous discussion offline, I fixed the getting of backlog value by considering extra factors of `flushTriggered` and `isFinished`, so it keeps the rule of updating backlog as before.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 1]
```diff
@@ -116,52 +115,58 @@ protected Throwable getFailureCause() {
...
 		}
-		return buffersInBacklog;
 	}
 
 	/**
 	 * Increases the number of non-event buffers by one after adding a non-event
 	 * buffer into this subpartition.
 	 */
-	protected void increaseBuffersInBacklog(BufferConsumer buffer) {
+	protected void increaseBuffersInBacklog(boolean isBuffer) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: This change should have been in a separate commit, since it's not related to the fix in any way, but just makes this method more consistent with `decreaseBuffersInBacklog` right?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java [hunk 2]
```diff
@@ -156,7 +156,8 @@ public BufferAndBacklog getNextBuffer() throws IOException, InterruptedException
 				checkState(nextBuffer.isFinished(),
 					"We can only read from SpillableSubpartition after it was finished");
 
-				newBacklog = parent.decreaseBuffersInBacklogUnsafe(nextBuffer.isBuffer());
+				parent.decreaseBuffersInBacklog(nextBuffer.isBuffer());
+				newBacklog = parent.getBuffersInBacklog();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java]: 1. why have you split this method into two? What was wrong with `decreaseBuffersInBacklog` returning the backlog?
2. why are you synchronizing here second time? You should be able to use `getBuffersInBacklogUnsafe()`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 3]
```diff
@@ -116,52 +115,58 @@ protected Throwable getFailureCause() {
...
+	/**
+	 * Gets the number of non-event buffers in this subpartition.
+	 */
+	public abstract int getBuffersInBacklog();
+
+	/**
+	 * @param lastBufferAvailable whether the last buffer in this subpartition is available for consumption
+	 * @return the number of non-event buffers in this subpartition
+	 */
+	protected int getBuffersInBacklog(boolean lastBufferAvailable) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: How is this method related to `PipelinedSubpartition#getNumberOfFinishedBuffers` and `PipelinedSubpartition#isAvailableUnsafe`. Can not they be unified/simplify in some way?  Or are they the same % that this `getBuffersInBacklog` counts only non-event buffers?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 4]
```diff
@@ -116,52 +115,58 @@ protected Throwable getFailureCause() {
...
+	}
+
 	/**
 	 * Decreases the number of non-event buffers by one after fetching a non-event
 	 * buffer from this subpartition (for access by the subpartition views).
-	 *
-	 * @return backlog after the operation
 	 */
-	public int decreaseBuffersInBacklog(Buffer buffer) {
+	public void decreaseBuffersInBacklogUnsafe(boolean isBuffer) {
```
[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: I misunderstood the semantic of `Unsafe` before. I define the `Unsafe` based on the external calling not internal implementation of `decreaseBuffersInBacklog`. In other words, if external method calls `decreaseBuffersInBacklog`, it should confirm the buffers are synchronized in the external method to call this safe, otherwise external method should call `decreaseBuffersInBacklogUnsafe`.

Now I understand your point of definition, and it seems more reasonable based on internal confirmation of `decreaseBuffersInBacklog`. I would revert this change as before.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Yes, I try to keep it consistent with another method. I should make it in a separate commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java (hunk 2 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java]: 1. The current backlog is based on `flushTriggered` or `finished` status after changing, if we want to return backlog during `decreaseBuffersInBacklog`, we might need to pass these info into `decreaseBuffersInBacklog(boolean `flushTriggered` | `finished`)`. In order not to dirty this method, I split it into a separate method. Maybe it does not seem so bad if passing parameter into `decreaseBuffersInBacklog `. So we could still merge these two into one method if you prefer.

2. Yes, you are right. Here we could get backlog in unsafe way, no need to add the synchronized overhead.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 3 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: That is very good question. We define the variable `buffersInBacklog` for counting the non-event buffers in the queue. `getNumberOfFinishedBuffers` and `isAvailableUnsafe` are counting both event and non-event buffers. So they can not be unified directly. I even tried another way by counting only event buffers instead of current `buffersInBacklog`, but it still does not simplify the logic. Especially for `SpillableSubpartition` we could not get total buffers directly from the memory queue, then we still need non-event buffer counting even though we count the event buffers.

I also find that the current conditions seem a bit complicated to maintain in different processes, such as `shouldNotifyDataAvailable`, `isAvailable`, `getBuffersInBacklog`, etc. And it would be better if we can integrate something to make related conditions unification. After we have a good idea for this issue, I would like to make changes. :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: I misunderstood the semantic of `Unsafe` before. I define the `Unsafe` based on the external calling not internal implementation of `decreaseBuffersInBacklog`. In other words, if external method calls `decreaseBuffersInBacklog`, it should confirm the buffers are synchronized in the external method to call this safe, otherwise external method should call `decreaseBuffersInBacklogUnsafe`.

Now I understand your point of definition, and it seems more reasonable based on internal confirmation of `decreaseBuffersInBacklog`. I would revert this change as before.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Yes, I try to keep it consistent with another method. I should make it in a separate commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java (hunk 2 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java]: 1. The current backlog is based on `flushTriggered` or `isFinished` status after changing, if we want to return backlog during `decreaseBuffersInBacklog`, we might need to pass these info into `decreaseBuffersInBacklog(boolean flushTriggered | isFinished)`. In order not to dirty this method, I split it into a separate method. Maybe it does not seem so bad if passing parameter into `decreaseBuffersInBacklog`. So we could still merge these two into one method if you prefer.

2. Yes, you are right. Here we could get backlog in unsafe way, no need to add the synchronized overhead.

**Comment by zhijiangW** [comment]
@pnowojski , thanks for your reviews and suggestions!

I would remove the second synchronized buffers during `getBuffersInBacklog`, so it would bring any regression in theory.  I agree with we should verify this changes would make some effects for certain scenarios. I would verify the current benchmark to confirm it. If no obvious improvement, I could adjust the scenarios to fit this change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java (hunk 2 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionView.java]: 1. Ok, I get it now :) Thanks for pointing this out.

I think that adding this extra parameter to the `decreaseBuffersInBacklog` would indeed be a worse solution and it's better to just use `getBuffersInBacklogUnsafe` here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 3 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Ok, this gets a little bit too complicated for my taste as well, but if you don't have idea how to fix it, lets keep it as it is :(

**Comment by zhijiangW** [comment]
@pnowojski , I just refactor the commit and update the codes.
Next I would verify the network benchmark and supplement the case if necessary.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 5]
```diff
@@ -118,13 +117,19 @@ protected Throwable getFailureCause() {
...
-	@VisibleForTesting
-	public int getBuffersInBacklog() {
-		return buffersInBacklog;
+	public abstract int getBuffersInBacklog();
+
+	/**
+	 * @param lastBufferAvailable whether the last buffer in this subpartition is available for consumption
+	 * @return the number of non-event buffers in this subpartition
+	 */
+	protected int getBuffersInBacklog(boolean lastBufferAvailable) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Hmm, I think I would keep the `Unsafe` suffix here, since this method is not synchronised compared to the `decreaseBuffersInBacklog`. Ditto for the abstract?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java [hunk 6]
```diff
@@ -147,7 +147,8 @@ public BufferAndBacklog getNextBuffer() throws IOException, InterruptedException
 			return null;
 		}
 
-		int newBacklog = parent.decreaseBuffersInBacklog(current.isBuffer());
+		parent.decreaseBuffersInBacklog(current.isBuffer());
+		int newBacklog = parent.getBuffersInBacklog();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java]: Something is wrong here. The line above is synchronized (`decreaseBuffersInBacklog` vs `decreaseBuffersInBacklogUnsafe`) and this `getBuffersInBacklog` is not (note miss-leading lack of `Unsafe` suffix) . Either:

1. the above one doesn't need to be synchronized
2. there is a bug and this one should be synchronized. In that case however, it would be additional unnecessary synchronisation cost, so I think I would prefer to integrate `getBuffersInBacklog` with `decreaseBuffersInBacklog` after all.
3. This doesn't need to be synchronized because of some hacky assumption. In that case either write big comment or probably better and/or safer approach would be to integrate `getBuffersInBacklog` with `decreaseBuffersInBacklog` after all.

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java]: That is a good question that should be concerned. I considered it again and thought the synchronized is not needed for getting backlog because this value would be final consistent between sender and receiver.

E.g. if the current backlog is 4 after decreasing, the previous behavior would report 4 strictly. The new behavior might report 4 or more than 4 if increasing backlog again before getting. But the result is still correct if reporting 5 because it actually exists. The difference is we report this increase in advance, and the previous behavior would reflect this increase in the next report. The early report might get some extra benefits because the receiver could prepare more credits for it.

It could have two options:
1. Uniform all the ways of getting backlog outside of synchronized.
2. Integrate return backlog in `decreaseBuffersInBacklog` as you mentioned, and the form seems not so bad if adding boolean parameter in `decreaseBuffersInBacklog`. The backlog is strictly returned in this way.

Which option do you prefer?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 5 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Yes, it makes sense to keep the `Unsafe` here consistent with `decreaseBuffersInBacklog`.

In another aspect, the existence of `Unsafe` is for distinguishing the different usages compared with `safe`. For `getBuffersInBacklog` there might need only one way for use either `safe` or `Unsafe`. Anyway I am willing to make it `Unsafe` here, then we could avoid explaining it in method comments.

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: I get your point, however the two usages of this method below showed that it's confusing as it is now. 
> Anyway I am willing to make it Unsafe here, then we could avoid explaining it in method comments.

Making code self explanatory is definitely a good thing :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java (hunk 6 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java]: I think that option 1. even if it's correct right now, would be more error prone for the future. Forcing every contributor/committer in this part to rethink this contract/dependencies every time is not a good thing to do.

For example, as you if the backlog increases between those two calls there is no problem, however it takes some time to realize about this fact. Also currently I'm not sure if the backlog can be somehow decreased by a 3rd party (another thread) betwen those two calls and if that happens, if that's ok? Another question, are all of the arithmetic operations on the backlog atomic? 

Long story short, if we can avoid think about the above issues, so much the better - I would pick number 2.

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java]: I agree with your point of option 2. Based on the current implementation we could confirm no problem, but it might leave risks if changing something in future. It is also easy to understand for others in option 2.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 7]
```diff
@@ -115,52 +115,62 @@ protected Throwable getFailureCause() {
 
 	public abstract boolean isReleased();
 
+	public abstract boolean isFinished();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Does it have to be public? If yes, then I think it should be named `isFinishedUnsafe()`. More or less we are creating a convention here as we go, but I think for private/protected methods we can ignore unsafe suffixes as long as there exists only one version of such method. Here existence of thread-safe `isReleased()` adds to the confusion and suggests that `isFinished()` should be thread-safe as well.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java [hunk 8]
```diff
@@ -115,52 +115,62 @@ protected Throwable getFailureCause() {
...
+
+	public abstract boolean isFlushRequested();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: The same question as above. Can it be non public or mark it unsafe?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 7 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: The access modifiers for these two methods could be narrowed down to `protected`. And I agree with ignoring the unsafe suffixes for them after changing to `protected`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java (hunk 8 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java]: Could be `protected` as above.

**Comment by zhijiangW** [comment]
Thanks for review again @pnowojski . I submitted the fixup commit for addressing your concern. :)

**Comment by zhijiangW** [comment]
@pnowojski  Yes, I would verify the benchmark results and extend the cases for idle channels. Maybe a bit delay for this work because some shuffle service works are also in parallel. I would focus on it this week. :)

**Comment by zhijiangW** [comment]
Thanks for reviews and merging @pnowojski !

As we discussed offline, this PR changes the backlog semantic from the number of producer buffers to the number of consumable buffers. In common sense I think this new semantic is more fit for the initial thought of introducing backlog for credit-based.

Your concern is reasonable.  As we discussed offline, it is hard to determine the specific effects for reporting high backlog size in different cases. If the extra requested floating buffers for high backlog are actually used soon on consumer side, it might have no effects. Otherwise it might bring negative effects because of floating buffers misdistribution. I agree with that there should have some follow-up works for further verifying the conclusion.

**Comment by pnowojski** [comment]
@zhijiangW can you create a follow up ticket to create such a test/benchmark? Also can you ping me once this is green, so I can merge it?

**Comment by zhijiangW** [comment]
Yes, I created the ticket [FLINK-13126](https://issues.apache.org/jira/browse/FLINK-13126) as a follow-up work. I guess the travis for this PR might not be finished today based on the current PR queues.

**Comment by zhijiangW** [comment]
@pnowojski  the travis is green.

**Comment by pnowojski** [comment]
Thanks, merged :)
